add contexts image other updates (#4262)
circleci/circleci-docs: jekyll/_cci2/contexts.md

order: 41
---

Contexts provide a mechanism for securing and sharing environment variables across projects. The environment variables are defined as name/value pairs and are injected at runtime. This document describes creating and using contexts in CircleCI in the following sections:

* TOC
{:toc}

## Overview
{:.no_toc}

> Create and manage contexts on the Organization Settings page of the CircleCI application. You must be an organization member to view, create, or edit contexts. After a context has been created, you can use the `context` key in the workflows section of a project [`config.yml`]({{ site.baseurl }}/2.0/configuration-reference/#context) file, to give any job(s) access to the environment variables associated with the context, as shown in the image below.

Author: ```suggestion
Create and manage contexts on the Organization Settings page of the CircleCI application. You must be an organization member to view, create, or edit contexts. After a context has been created, you can use the `context` key in the workflows section of a project [`config.yml`]({{ site.baseurl }}/2.0/configuration-reference/#context) file to give any job(s) access to the environment variables associated with the context, as shown in the image below.
```

Organization administrators have read/write access to all projects and have unrestricted access to all contexts.

The default security group is `All members` and enables any member of the organization who uses CircleCI to use the context.

**Note:** Bitbucket repositories do **not** provide an API that allows CircleCI contexts to be restricted, only GitHub projects include the ability to restrict contexts with security groups.

## Running Workflows with a Restricted Context

To invoke a job that uses a restricted context, a user must be a member of one of the security groups for the context. If the user running the workflow does not have access to the context, the workflow will fail with the `Unauthorized` status.

### Restrict a Context to a Security Group or Groups
{:.no_toc}

You must be an organization administrator to complete the following task.

1. Navigate to Organization Settings > Contexts page in the CircleCI app. The list of contexts appears. The default security group is `All members` and allows all users in the org to invoke jobs with that context.
2. Click the Create Context button if you wish to use a new context, or click the name of an existing context.

> 3. Click the Add Secutrity Group button to view the dialog bo.

Reviewer: ```suggestion
3. Click the Add Security Group button to view the dialog box.
```

Author: wowzers you can tell my head is in two places at once, thanks Thomas!

3. Click the Add Security Group button to view the dialog box.
4. Select GitHub teams or LDAP groups to add to the context and click the Add button. Use of the context is now limited to the selected groups.
5. Click Add Environment Variables to add environment variables to the context if none exist and click the Add button. Use of the environment variables for this context is now limited to members of the security groups.
6. Navigate to Organization Settings > Contexts in the CircleCI app. The security groups appear in the Security column for the context.

Only members of the selected groups may now use the context in their workflows or add or remove environment variables for the context. 

### Approving Jobs that use Restricted Contexts
{:.no_toc}

Adding an [approval job]({{ site.baseurl }}/2.0/configuration-reference/#type) to a workflow gives the option to require manual approval of the use of a restricted context. To restrict running of jobs that are downstream from an approval job, add a restricted context to those downstream jobs, as shown in the example below:

```yaml

> workflows:

Reviewer 2: As this is a partial example (not displaying the jobs that are referenced) I would maybe add a commend above the workflows that indicates other text is expected to exist (`# ...`)

Author: Oh yes good idea 👍

3. Click the Add Security Group button to view the dialog box.
4. Select GitHub teams or LDAP groups to add to the context and click the Add button. Use of the context is now limited to the selected groups.
5. Click Add Environment Variables to add environment variables to the context if none exist and click the Add button. Use of the environment variables for this context is now limited to members of the security groups.
6. Navigate to Organization Settings > Contexts in the CircleCI app. The security groups appear in the Security column for the context.

Only members of the selected groups may now use the context in their workflows or add or remove environment variables for the context. 

### Approving Jobs that use Restricted Contexts
{:.no_toc}

Adding an [approval job]({{ site.baseurl }}/2.0/configuration-reference/#type) to a workflow gives the option to require manual approval of the use of a restricted context. To restrict running of jobs that are downstream from an approval job, add a restricted context to those downstream jobs, as shown in the example below:

```yaml
version: 2.1

> # jobs declaration from build, test and deploy not displayed for brevity

Author: ```suggestion
# jobs declaration for build, test and deploy not displayed
```
